pkg/metrics_store/metrics_writer.go

```diff
@@ -82,13 +87,22 @@ func (m MetricsWriter) WriteAll(w io.Writer) error {
 	return nil
 }
 
-// SanitizeHeaders removes duplicate headers from the given MetricsWriterList for the same family (generated through CRS).
-// These are expected to be consecutive since G** resolution generates groups of similar metrics with same headers before moving onto the next G** spec in the CRS configuration.
-func SanitizeHeaders(writers MetricsWriterList) MetricsWriterList {
+// SanitizeHeaders sanitizes the headers of the given MetricsWriterList.
+func SanitizeHeaders(contentType string, writers MetricsWriterList) MetricsWriterList {
 	var lastHeader string
 	for _, writer := range writers {
 		if len(writer.stores) > 0 {
 			for i, header := range writer.stores[0].headers {
+				// If the requested content type was proto-based, replace the type with "gauge", as "info" and "statesets" are not recognized by Prometheus' protobuf machinery.
+				if strings.HasPrefix(contentType, expfmt.ProtoFmt) &&
+					strings.HasPrefix(header, "# HELP") &&
```

Comment:
Have you considered using a regular expression here?

Reply:
I was on the fence between using a regex and the in-built `strings` for this, since the latter seemed more readable, and the exposition format isn't prone to change. But LMK and I'll revert.

Reply:
I was mainly wondering how costly the operation is. Might be worth looking at our benchmarks in CI

Reply:
Benchmarking `regex` v/s `strings` showed the following results.

### `regex` snippet

```go
if strings.HasPrefix(contentType, expfmt.ProtoType) {
	pattern := `^# HELP.*\n# TYPE.* (info|stateset)`
	matched, err := regexp.MatchString(pattern, header)

	// Fail silently.
	if err == nil && matched {
		typeStringWithoutTypePaddedIndex := strings.LastIndex(header, " ")
		typeStringWithoutType := header[:typeStringWithoutTypePaddedIndex]
		writer.stores[0].headers[i] = typeStringWithoutType + " " + string(metric.Gauge)
	}
}
```

### `strings` snippet

```go
if strings.HasPrefix(contentType, expfmt.ProtoType) &&
	strings.HasPrefix(header, "# HELP") &&
	(strings.HasSuffix(header, " "+string(metric.Info)) || strings.HasSuffix(header, " "+string(metric.StateSet))) {
	typeStringWithoutTypePaddedIndex := strings.LastIndex(header, " ")
	typeStringWithoutType := header[:typeStringWithoutTypePaddedIndex]
	writer.stores[0].headers[i] = typeStringWithoutType + " " + string(metric.Gauge)
}
```

### Benchmarks

`strings` seems to be ~95% faster than `regex`.

```
┌[rexagod@nebuchadnezzar] [/dev/ttys003] [2248 ⚡]
└[~/repositories/oss/kube-state-metrics]> benchstat regex.txt strings.txt
goos: darwin
goarch: arm64
pkg: k8s.io/kube-state-metrics/v2/pkg/metrics_store
                                                  │   regex.txt    │             strings.txt             │
                                                  │     sec/op     │   sec/op     vs base                │
SanitizeHeaders/text-format_unique_headers-10          559.2µ ± 0%   560.7µ ± 0%   +0.27% (p=0.000 n=10)
SanitizeHeaders/text-format_duplicate_headers-10       528.2µ ± 0%   528.6µ ± 1%   +0.08% (p=0.035 n=10)
SanitizeHeaders/proto-format_unique_headers-10       388.639m ± 2%   3.981m ± 0%  -98.98% (p=0.000 n=10)
SanitizeHeaders/proto-format_duplicate_headers-10   356906.4µ ± 0%   590.2µ ± 0%  -99.83% (p=0.000 n=10)
geomean                                                14.23m        913.5µ       -93.58%

                                                  │   regex.txt    │               strings.txt               │
                                                  │      B/op      │    B/op     vs base                     │
SanitizeHeaders/text-format_unique_headers-10         0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
SanitizeHeaders/text-format_duplicate_headers-10      0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10) ¹
SanitizeHeaders/proto-format_unique_headers-10      737.8Mi ± 0%     0.0Mi ± 0%  -100.00% (p=0.000 n=10)
SanitizeHeaders/proto-format_duplicate_headers-10   733.2Mi ± 0%     0.0Mi ± 0%  -100.00% (p=0.000 n=10)
geomean                                                          ²               ?                       ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean

                                                  │   regex.txt   │               strings.txt                │
                                                  │   allocs/op   │  allocs/op   vs base                     │
SanitizeHeaders/text-format_unique_headers-10        0.000 ± 0%      0.000 ± 0%         ~ (p=1.000 n=10) ¹
SanitizeHeaders/text-format_duplicate_headers-10     0.000 ± 0%      0.000 ± 0%         ~ (p=1.000 n=10) ¹
SanitizeHeaders/proto-format_unique_headers-10      4.101M ± 0%     0.000M ± 0%  -100.00% (p=0.000 n=10)
SanitizeHeaders/proto-format_duplicate_headers-10   4.100M ± 0%     0.000M ± 0%  -100.00% (p=0.000 n=10)
geomean                                                         ²                ?                       ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean
```

The benchmarking suite used is present [here](https://github.com/kubernetes/kube-state-metrics/pull/2270/commits/640960336ac58d655a83f81ccbe789bbced58c3f#diff-a2fba22b7063f6a5f98f7c677327f9d87bc5f07d84cd567009a1f41ce6799adeR268).